posthog/api/prompt.py

```diff
@@ -1,93 +1,231 @@
-                if saved_state and local_state["last_updated_at"] > saved_state.last_updated_at:
-                    saved_state.last_updated_at = local_state["last_updated_at"]
-                    saved_state.step = local_state["step"]
-                    saved_state.completed = local_state.get("completed", False)
-                    saved_state.dismissed = local_state.get("dismissed", False)
+                if saved_state and local_state.last_updated_at > saved_state.last_updated_at:
+                    saved_state.last_updated_at = local_state.last_updated_at
+                    saved_state.step = local_state.step
+                    saved_state.completed = local_state.completed
+                    saved_state.dismissed = local_state.dismissed
                     states_to_update.append(saved_state)
+                    state = saved_state
+                elif saved_state is not None:
+                    state = saved_state
+                else:
+                    states_to_create.append(local_state)
                     state = local_state
-                elif saved_state is None:
-                    state = local_state
-                    new_state = UserPromptSequenceState(user=request.user, **local_state)
-                    states_to_create.append(new_state)
-
-            if not state and saved_state:
-                state = PromptSequenceStateSerializer(saved_state).data
+            else:
+                if saved_state:
+                    state = saved_state
+                # if the sequence should autorun for all users, we create a state with no step, meaning the user has not seen it but should start seeing it
+                elif sequence.autorun:
+                    state = UserPromptState(user=request.user, sequence=sequence, step=None)
 
             if state:
-                new_states.append(state)
-                my_prompts["state"][sequence["key"]] = state
+                up_to_date_states.append(state)
 
+        my_prompts: Dict[str, Any] = {"state": {}, "sequences": []}
         # filter only the sequences where `must_be_completed` rule passes
-        sequences = [x for x in all_sequences if not x["rule"].get("must_be_completed")]
-        sequences_requiring_previous_completion = [x for x in all_sequences if x["rule"].get("must_be_completed")]
-
-        for seq in sequences_requiring_previous_completion:
-            must_be_completed = seq["rule"]["must_be_completed"]
-            current_state: Optional[Dict] = next((s for s in new_states if s["key"] in must_be_completed), None)
-            if not current_state or not current_state["completed"]:
-                continue
-            sequences.append(seq)
-
-        my_prompts["sequences"] = sequences
-
+        # this allows to run certain sequences only if other have been completed first
+        for state in up_to_date_states:
+            sequence = state.sequence
+            must_have_completed = sequence.must_have_completed.all()
```

Comment:
For reviewers, since a sequence has now a `ManyToMany` field to check for sequences that must have completed first, we call here `must_have_completed.all()`, but I'm not sure if we can do any optimization to avoid this